src/main/java/org/sonarlint/intellij/fs/EditorFileChangeListener.kt

```diff
@@ -19,26 +19,69 @@
  */
 package org.sonarlint.intellij.fs
 
+import com.google.common.annotations.VisibleForTesting
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.editor.EditorFactory
 import com.intellij.openapi.editor.event.BulkAwareDocumentListener
-import com.intellij.openapi.editor.event.DocumentEvent
 import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.openapi.module.Module
 import com.intellij.openapi.project.Project
+import com.intellij.openapi.startup.StartupActivity
 import org.sonarlint.intellij.core.ProjectBindingManager
 import org.sonarlint.intellij.util.SonarLintAppUtils
 import org.sonarlint.intellij.util.SonarLintUtils.getService
+import org.sonarsource.sonarlint.core.client.api.common.ClientModuleFileEvent
 import org.sonarsource.sonarlint.plugin.api.module.file.ModuleFileEvent
+import java.util.Timer
+import java.util.TimerTask
 
-class EditorFileChangeListener(
-    private val project: Project,
-    private val fileEventsNotifier: ModuleFileEventsNotifier = getService(
-        ModuleFileEventsNotifier::class.java
-    )
-) : BulkAwareDocumentListener {
-    override fun documentChanged(event: DocumentEvent) {
-        val engine = getService(project, ProjectBindingManager::class.java).engineIfStarted ?: return
-        val file = FileDocumentManager.getInstance().getFile(event.document) ?: return
+const val DEBOUNCE_DELAY_MS = 1000L
+
+class EditorFileChangeListener : BulkAwareDocumentListener.Simple, StartupActivity {
+
+    private var fileEventsNotifier: ModuleFileEventsNotifier? = null
+    private lateinit var project: Project
+    private var eventsToSend = LinkedHashMap<Module, MutableMap<String, ClientModuleFileEvent>>()
+    private var triggerTimer = initTimer()
```

Comment:
Not too sure but as you create a timer instance, it might be better to dispose it when the project closes. Maybe it's sufficient to implement Disposable here and stop the timer, but not sure it would be called. Another option is to call `Disposer.register` with the project as the parent